dev/com.ibm.ws.concurrent.mp_fat/test-applications/MPConcurrentTxApp/src/concurrent/mp/fat/tx/web/MPConcurrentTxTestServlet.java

```diff
@@ -0,0 +1,225 @@
+                    .build();
+
+    private ManagedExecutor txExecutor = ManagedExecutor.builder()
+                    .propagated(ThreadContext.TRANSACTION)
+                    .cleared(ThreadContext.ALL_REMAINING)
+                    .build();
+
+    @Override
+    public void destroy() {
+        testThreads.shutdownNow();
+    }
+
+    @Override
+    public void init(ServletConfig config) throws ServletException {
+        testThreads = Executors.newFixedThreadPool(5);
+
+        // create tables for tests to use and pre-populate each with a single entry
+        try {
+            try (Connection con = defaultDataSource.getConnection(); Statement st = con.createStatement()) {
+                st.execute("CREATE TABLE MNCOUNTIES (NAME VARCHAR(50) NOT NULL PRIMARY KEY, POPULATION INT NOT NULL)");
+                st.execute("CREATE TABLE IACOUNTIES (NAME VARCHAR(50) NOT NULL PRIMARY KEY, POPULATION INT NOT NULL)");
+                st.execute("INSERT INTO MNCOUNTIES VALUES ('Olmsted', 154930)");
+                st.execute("INSERT INTO IACOUNTIES VALUES ('Polk', 481830)");
+            }
+        } catch (SQLException x) {
+            throw new ServletException(x);
+        }
+    }
+
+    @Test
+    public void testTransactionPropagatedToSameThread() throws Exception {
+        CompletableFuture<Integer> stage0 = txExecutor.newIncompleteFuture();
+        CompletableFuture<Integer> stage1, stage2, stage3;
+        tx.begin();
+        try {
+            stage1 = stage0.thenApply(numUpdates -> {
+                try (Connection con = defaultDataSource.getConnection(); Statement st = con.createStatement()) {
+                    return numUpdates + st.executeUpdate("INSERT INTO MNCOUNTIES VALUES ('Winona', 50992)");
+                } catch (Exception x) {
+                    throw new CompletionException(x);
+                }
+            });
+            stage2 = stage1.thenApply(numUpdates -> {
+                try (Connection con = defaultDataSource.getConnection(); Statement st = con.createStatement()) {
+                    return numUpdates + st.executeUpdate("INSERT INTO MNCOUNTIES VALUES ('Rice', 65251)");
+                } catch (Exception x) {
+                    throw new CompletionException(x);
+                }
+            });
+            stage3 = stage2.thenApply(numUpdates -> {
+                try (Connection con = defaultDataSource.getConnection(); Statement st = con.createStatement()) {
+                    numUpdates += st.executeUpdate("INSERT INTO MNCOUNTIES VALUES ('Fillmore', 20825)");
+                    tx.commit();
+                    return numUpdates;
+                } catch (Exception x) {
+                    throw new CompletionException(x);
+                }
+            }).exceptionally(x -> {
+                try {
+                    tx.rollback();
+                } catch (Exception x2) {
+                    throw new CompletionException(x);
```

Comment:
We could consider condensing both "throw new CompletionException(x);" statements into a single finally block.  Did you intend to ignore the occurrence of an exception during rollback and rethrow the original cause of the problem?

Reply:
Yes, I did intend to raise the original exception, but that is the initial cause and the one that matters. I'll at least log x2 to System.out in case anyone wants it for extra debug though.